Add in jemalloc glibc hooking again (#2591)
FEX-Emu/FEX: docs/jemalloc_usage.md

# JEMalloc usage

> FEX-Emu uses two different jemalloc heap allocators at once.
> - jemalloc: The primary heap allocator
> - jemalloc_glibc: The second heap allocator

Reviewer: ```suggestion
FEX-Emu uses two different jemalloc heap allocators at once, each for different purposes:
- jemalloc: The primary heap allocator (to keep FEX's internal allocations out of the 32-bit address space used by guest applications)
- jemalloc_glibc: The second heap allocator (to add allocation introspection features used by thunks)
```

Author: Changed phrasing to yours.

## jemalloc - primary heap allocator

> This heap allocator is the primary allocator that FEX-Emu uses. All of FEXCore's `fextl::` namespaced objects will allocate memory through this
> allocator. This allocator has its usage of mmap and munmap overridden to use FEXCore's internal VMA region allocator.

Reviewer: Phrasing is a bit roundabout here. I'd describe what this actually does first, then where it's used (since the section header already mentions it's the primary allocator):

```suggestion
This allocator overrides `mmap` and `munmap` by forwarding them to FEXCore's internal VMA region allocator.

All of FEXCore's `fextl::` namespaced objects allocate memory with this method.
```

Author: Changed phrasing to yours.

### FEXCore internal VMA region allocator

> - When running a 64-bit guest application, this VMA region allocator is disabled and passes through to mmap and munmap in the host kernel.
> - This only gets interesting when running a 32-bit guest application.
> 
> When running a 32-bit guest application, the VMA region allocator will allocate all memory above 4GB virtual address space (Up to 256TB on 48-bit VA
> systems!). It then provides a mmap and munmap interface to allow allocating memory out of this reserved virtual address space. This ensures that all
> of FEX's allocations stay out of the lower 32-bit 4GB VA space, otherwise games would quickly run out of virtual address space.

Reviewer: This is backwards: Instead of starting with the case where literally nothing happens, start with the interesting bit:

```suggestion
When running a 32-bit guest application, the VMA region allocator allocates from memory *above* the 4GB of virtual address space reserved for the 32-bit application.

This ensures that all of FEX's allocations stay out of the lower 32-bit 4GB VA space, since games would quickly run out of virtual address space otherwise.

When running a 64-bit guest application, this VMA region allocator is disabled and passes through to mmap and munmap in the host kernel.
```

> (Up to 256TB on 48-bit VA systems!)

Since we're describing fairly intricate logic here, omitting details like this makes it easier to digest the important pieces.

> It provides a mmap and munmap interface to allow allocating memory out of this reserved virtual address space.

This seems redundant. It's an allocator, so of course it provides those interfaces.

Author: Changed phrasing to yours.

## jemalloc_glibc - secondary heap allocator

> This heap allocator replaces glibc's allocator using weak symbol overriding. This explicitly does **not** use the FEXCore VMA region allocator. This

Reviewer: > This explicitly does **not** use the FEXCore VMA region allocator

This is backwards again: It's starting to explain something by explaining what it's not.

> This heap allocator replaces glibc's allocator using weak symbol overriding. This explicitly does **not** use the FEXCore VMA region allocator. This
> ensures that all allocations that go through this heap allocator will use the kernel mmap and munmap interface. All shared libraries will use this
> allocator including the host-side of thunks.

Reviewer: > This explicitly does not use the FEXCore VMA region allocator

This is backwards again: It's starting to explain something by explaining what it's not.

> This ensures that all allocations that go through this heap allocator will use the kernel mmap and munmap interface.

This describes the technical implementation but not what the allocator intends to achieve.

```suggestion
This heap allocator replaces the host glibc's allocator using weak symbol overriding. It adds introspection features used for thunking, but has no functional differences otherwise: All memory is allocated in the 4GB of 32-bit address space. The FEXCore VMA region allocator is explicitly **not** involved hence.

All native shared libraries use this allocator including the host-side of thunks.

Internally, all allocations that go through this heap allocator use the kernel mmap and munmap interface.
```

Author: Changed phrasing to yours.

### Thunks

> When thunks allocate memory, some pointers need to be tracked from which heap allocator they come from when getting freed. In this case some pointers
> might be coming from the guest heap-allocator, and some might be coming from the host heap-allocator.
> 
> jemalloc provides a `is_known_allocation` function that allows us to query if the pointer comes from the host heap-allocator. Which FEXCore can use to
> determine which heap should free the pointer. glibc doesn't provide this same interface which is why we need to use jemalloc for this.

Reviewer: ```suggestion
Thunks may allocate memory either through the guest-side (on the guest glibc heap) or the host-side (on the `jemalloc_glibc` heap). To properly free this memory, FEX must be able to determine which heap allocator it belongs to.

`glibc` provides no public interface to do this, but FEX's `jemalloc` fork does. The `is_known_allocation` function is used by FEXCore to query whether a given pointer originated from the `jemalloc_glibc` allocator. This enables FEXCore to determine the appropriate heap for freeing the pointer.
```

(`verify`->`query`, since the former suggests the condition is expected to hold true)

Author: Changed phrasing to yours.
